cmd/clusterctl/cmd/version_checker.go

```
@@ -0,0 +1,204 @@
+func (v *versionChecker) Check() (string, error) {
+	log := logf.Log
+	cliVer, err := semver.ParseTolerant(v.cliVersion().GitVersion)
+	if err != nil {
+		return "", errors.Wrap(err, "unable to semver parse clusterctl GitVersion")
+	}
+
+	release, err := v.getLatestRelease()
+	if err != nil {
+		return "", err
+	}
+	if release == nil {
+		return "", nil
+	}
+	latestVersion, err := semver.ParseTolerant(release.Version)
+	if err != nil {
+		return "", errors.Wrap(err, "unable to semver parse latest release version")
+	}
+
+	// if we are using a dirty dev build, just log it out
+	if strings.HasSuffix(cliVer.String(), "-dirty") {
+		log.V(1).Info("⚠️ Using a development build of clusterctl.", "CLIVersion", cliVer.String(), "LatestGithubRelease", release.Version)
+		return "", nil
+	}
+
+	// if the cli version is a dev build off of the latest available release,
+	// the just log it out as informational.
+	if strings.HasPrefix(cliVer.String(), latestVersion.String()) && gitVersionRegEx.MatchString(cliVer.String()) {
+		log.V(1).Info("⚠️ Using a development build of clusterctl.", "CLIVersion", cliVer.String(), "LatestGithubRelease", release.Version)
+		return "", nil
+	}
+
+	if cliVer.Major == latestVersion.Major &&
+		cliVer.Minor == latestVersion.Minor &&
+		latestVersion.GT(cliVer) {
+		return fmt.Sprintf(`
+New clusterctl version available: v%s -> v%s
+%s
+`, cliVer, latestVersion.String(), release.URL), nil
+	}
+	return "", nil
+}
+
+func (v *versionChecker) getLatestRelease() (*ReleaseInfo, error) {
+	log := logf.Log
+	vs, err := readStateFile(v.versionFilePath)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to read version state file")
+	}
+
+	// if there is no release info in the state file, pull latest release from github
+	if vs == nil {
+		release, _, err := v.githubClient.Repositories.GetLatestRelease(context.TODO(), "kubernetes-sigs", "cluster-api")
+		if err != nil {
+			log.V(1).Info("⚠️ Unable to get latest github release for clusterctl")
```

Are we avoiding v(0) to avoid printing logs in the middle of yaml output? otherwise I think it makes sense to print this as a warning on all runs

---

This was in regards to a previous feedback comment about failing silently. So I thought putting it at a log level that's informative only when asked for. As per the [logging guidelines laid out here](https://github.com/kubernetes-sigs/cluster-api/blob/7e445a129abb28b20ede1040cc64f8766059eab2/cmd/clusterctl/log/doc.go#L43).

Also since this check is being run after every command, I didn't want the user to be peppered with the log line all the time.